src/interface/report/BossEventLoader.js

```diff
@@ -0,0 +1,169 @@
+import PropTypes from 'prop-types';
+import { withRouter } from 'react-router-dom';
+import { Trans } from '@lingui/macro';
+import ExtendableError from 'es6-error';
+
+import { fetchEvents, LogNotFoundError } from 'common/fetchWclApi';
+import { captureException } from 'common/errorLogger';
+import makeAnalyzerUrl from 'interface/common/makeAnalyzerUrl';
+import { findByBossId } from 'raids';
+
+import Ghuun from './images/Ghuun.gif';
+import handleApiError from './handleApiError';
+import './EventParser.css';
+
+export class BossEventsLoadError extends ExtendableError {
+  reason = null;
+  constructor(reason) {
+    super();
+    this.reason = reason;
+  }
+}
+
+class BossEventsLoader extends React.PureComponent {
+  static propTypes = {
+    report: PropTypes.shape({
+      title: PropTypes.string.isRequired,
+      code: PropTypes.string.isRequired,
+    }).isRequired,
+    fight: PropTypes.shape({
+      start_time: PropTypes.number.isRequired,
+      end_time: PropTypes.number.isRequired,
+      boss: PropTypes.number.isRequired,
+    }).isRequired,
+    children: PropTypes.func.isRequired,
+    history: PropTypes.shape({
+      push: PropTypes.func.isRequired, // adds to browser history
+      replace: PropTypes.func.isRequired, // updates current browser history entry
+      location: PropTypes.shape({
+        pathname: PropTypes.string.isRequired,
+      }).isRequired,
+    }),
+  };
+  state = {
+    error: null,
+    finished: false,
+    bossEvents: null,
+  };
+
+  componentDidMount() {
+    // noinspection JSIgnoredPromiseFromCall
+    this.parse();
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    const changed = this.props.report !== prevProps.report
+      || this.props.fight !== prevProps.fight;
+    if (changed) {
+      // noinspection JSIgnoredPromiseFromCall
+      this.parse();
+    }
+  }
+
+  async parse() {
+    let events = null;
+    return Promise.all([
+      this.loadEvents().then(result => {
+        events = result;
+      }),
+    ])
+      .then(() => {
+        return this.updateEvents(events);
+      })
+      .catch(error => {
+        const isCommonError = error instanceof LogNotFoundError;
+        if (!isCommonError) {
+          captureException(error);
+        }
+        this.reset();
+        this.setState({
+          error,
+        });
+      });
+
+  }
+  async updateEvents(events) {
+    const updatedEvents = events.map(e => ({
+      ...e,
+      isBossEvent: true,
+    }));
+
+    this.setState({bossEvents: updatedEvents, finished: true});
+  }
+
+  loadEvents() {
+    const { report, fight } = this.props;
+
+    const bossConfig = findByBossId(fight.boss);
+    if (bossConfig.fight.phases && bossConfig.fight.phases.length !== 0) {      
+      const filters = [];
+
+      const phasesForDifficulty = bossConfig.fight.phases.filter(phase => phase.difficulties.includes(fight.difficulty));
+
+      const abilities = phasesForDifficulty.reduce((abilityIds, phase) => {
```

Comment:
Why do abilities get special treatment? I don't understand the way you're building the filter; shouldn't the abilities be a part of the filter they're listed with instead of being a global filter? The global filter seems like it might restrict what we can make for filters.

Reply:
What do you mean here? I'm trying to group all of the filters as best as possible to simplify the query logic for WCL.

Reply:
Doesn't it do `(type='removebuff' and ability.id=263504)`?

Reply:
Nope. I get all the different ability ids from the config that trigger phase changes, and get all events based with that id. I then filter those events during my normalizer at the moment. All the abilities that have been used to trigger events so far have very few events associated with them so I'm not having loads of useless events passed back

Reply:
Is there any reason not to combine it? You could make a filter builder as simple as:

```jsx
export default function makeWclFilter(obj) {
  const filter = [];
  Object.keys(obj).forEach(key => {
    const value = obj[key];
    if (typeof value === 'number') {
      filter.push(`${key}=${value}`);
    } else {
      filter.push(`${key}='${value}'`);
    }
  });
  return filter.join(' AND ');
}

```

Reply:
I could certainly extract my function out, but I feel that my version does produce a better filter.

(ability.id IN (1, 2, 3, 4, 5, x)) 
OR 
(Custom_query_1) // used for add wave
OR
(Custom_query_2) // used for different add wave

Reply:
Why is it more optimised? This would get several unnecessary events if the spell were used for other things.

Reply:
Instead of having ability.id = 1 or ability.id = 2 or ability.id = 3